### DISCUSSION TITLE
Making templates modular

### DISCUSSION THREAD

**Discussion by whillas-yabble** [copier-org/copier#1874] [Ideas]
So my frustration with cookiecutter is that every project repo is very opinionated and at the same time incomplete. Your CI/CD or cloud provider might not be provided for in the template but the best practices are well know and evolving. 

Take requirements.txt vs pyproject.toml, you might want to use one over the other (or nether) depending on legacy requirements or not. You want to use which ever one you choose over all your application types i.e. django, fastapi, plain python module, machine learning etc. If the approach changes you want you next template based project generation to just start using the new change when you run copier e.g. Copier config refers to a "requirements-template" git repo which is maintained separately from your own and just has the most modern approach to requirements management and nothing else and may only have one or two files that it generates.

or another example

which package manage your using i.e. vanilla pip, pipenv, poetry, uv, something propriety etc. Expecting every template to keep up to date across all of these is a lot of work and just going to go out of date quickly. Or you might want change from one to another and get all the latest best practices by just referring to a 3rd party maintained git repo.

Same for cloud (or no cloud), documentation generation, django project layout etc

I guess the idea is that the templates people make are like best practices, which are constantly changing, evolving and branching. It would be more manageable if they were modular and could be composed together. 

I feel like this is the next major step in the project template generation game.

Just thinking out loud...

**Comment by whillas-yabble** [comment]
I guess one might try and use `copier.run_copy()` in a script which could pull multiple repos with the same destination folder. But this is imperative not declarative :(

**Reply by whillas-yabble** [reply]
I guess that would be an easy fix, just read in yaml. Something like
```yaml
project:
   templates:
   - github.com/person-a/coder-requirements-template
   - github.com/person-b/coder-uv-template@v1.1
   - github.com/person-c/coder-cutting-edge-django-template@master
   - etc
 ```
Just pull and run them in order I guess?
Latter ones override the previous ones if there are file conflicts?
Be good if they could share a file, i.e. `README.md` or `pyproject.toml`, and append instead of overwrite?

**Comment by yajo** [comment]
If each copier template targets a different set of files, you can have various templates applied in the same subproject and update each one individually.

**Comment by markdoerr** [comment]
I think, the proposals of @whillas-yabble and @yajo,
should be explored in a POC implementation.

I had a similar need and created an issue for that: #2261

**Comment by markdoerr** [comment]
@sisp pointed out, there is a long discussion thread in #934.
